Replace cloud provider forms with an API-driven DDF approach

skateman opened the pull request:
Depends on: https://github.com/ManageIQ/manageiq-api/pull/741 https://github.com/ManageIQ/manageiq-api/pull/830
Replaces: https://github.com/manageiq/manageiq-ui-classic/pull/6402
Parent issue: https://github.com/ManageIQ/manageiq/issues/18818

The following providers are already supported:
* [x] Amazon - https://github.com/ManageIQ/manageiq-providers-amazon/pull/599
* [x] GCE
* [x] OpenStack - https://github.com/ManageIQ/manageiq-providers-openstack/pull/568
* [x] vCloud - https://github.com/ManageIQ/manageiq-providers-vmware/pull/561
* [x] Azure
* [x] Azure stack - https://github.com/ManageIQ/manageiq-providers-azure_stack/pull/26

Known issues:
* [x] Initial value loading when changing between types - https://github.com/data-driven-forms/react-forms/issues/332
* [x] Initial value prioritization when editing - https://github.com/ManageIQ/manageiq-ui-classic/pull/6780
* [x] ~~Schema change for determining endpoint/authentication types~~ **NOT NECESSARY**

@miq-bot add_label react, enhancement, compute/cloud

skateman commented on app/javascript/forms/provider-forms/cloud-provider-form/index.jsx:
```diff
@@ -0,0 +1,170 @@
+      label: __('Type'),
+      placeholder: `<${__('Choose')}>`,
+      loadOptions: () =>
+        API.options('/api/providers').then(({ data: { supported_providers } }) => supported_providers // eslint-disable-line camelcase
```
That's a discussion outside of the context of this pr

himdel replied:
Let's move it to https://github.com/ManageIQ/manageiq-ui-classic/pull/6711 :)

skateman commented:
I'm sure I'm not able to write tests, so I will rely on the experts 😁 @brumik can you help me in a followup pr?

brumik commented:
I think so :)

skateman commented on app/javascript/forms/provider-forms/cloud-provider-form/index.jsx:
```diff
@@ -0,0 +1,170 @@
+  API.get('/api/zones?expand=resources&attributes=id,name,visible&filter[]=visible!=false&sort_by=name')
+    .then(({ resources }) => resources.map(({ name }) => ({ value: name, label: name })));
+
+const initialSchema = includeType => ([
```
Yeah, I'll replace it with a static object, no function needed. Originally the `type` field was not part of the form.

Hyperkid123 commented on app/javascript/forms/provider-forms/cloud-provider-form/index.jsx:
```diff
@@ -0,0 +1,170 @@
+  return (
+    <div>
+      { !providerId && (
+        <MiqFormRenderer
```
I am starting to suspect that the issue with the empty initial values may be caused by these two form instances which have the same field name. Can you try using the select without the renderer? It is exported in `rawComponents` object (i know stupid name..)

skateman replied:
I don't really understand what you're asking. But IMO having 2 form instances with the same field should be a valid workflow.

Hyperkid123 replied:
It is, and the state should be isolated, but can you just try it out?

skateman replied:
Can I just simply get rid of the first form for a test run?

Hyperkid123 replied:
yup

Hyperkid123 commented on app/javascript/forms/provider-forms/cloud-provider-form/index.jsx:
```diff
@@ -0,0 +1,170 @@
+          schema={schema}
+          onSubmit={onSubmit}
+          onCancel={onCancel}
+          initialValues={values}
```
Wait second. I know what it does not work. the initialValues is mutable, so you can't change it in runtime. Try unmnout the second form until you have the data based on the provider type.

Hyperkid123 replied:
In order to load the initial values, you would have to call `initialize` method form the `formOptions`. We will need a component for that.

Right now, the fields are not filed with values because the were initialized with empty values.

skateman replied:
This might be a problem, because you should be able to change the provider type (1st form) during runtime.

skateman replied:
Is it possible to force-remount the component if `values` change?

Hyperkid123 replied:
It is not possible nor the right approach. This will need a custom component that will change the values when the type is changed.

Hyperkid123 replied:
You can try `initializeOnMount` but it dont think it will help

Hyperkid123 replied:
@skateman ok this seems to be possible in a later version of the `final-form` API. We are currently in progress of upgrading to the latest version of final form, but bring some breaking changes. We can test it when we have a pre-release out. That should not take long.

skateman replied:
I can live with that.

Hyperkid123 replied:
I will do some more testing to confirm that the upgrades allow form/field level mutation of initial values.

himdel commented on app/javascript/forms/provider-forms/cloud-provider-form/index.jsx:
```diff
@@ -0,0 +1,170 @@
+
+const loadProviderServerZones = () =>
+  API.get('/api/zones?expand=resources&attributes=id,name,visible&filter[]=visible!=false&sort_by=name')
+    .then(({ resources }) => resources.map(({ name }) => ({ value: name, label: name })));
```
Oh, this bit sounds wrong,
should we really be selecting zones based on name not id?

Hyperkid123 replied:
Apparently that is what API wants. There was ID before.

skateman replied:
Well, it IS wrong, but this is what the API wants :laughing:

himdel replied:
Ah :D :+1:

himdel replied:
Though, in that case, we probably don't need `attributes=id,visible`

skateman commented:
I was able to fix the loading of initial values, but it required some serious changes in the form. After these changes the commits didn't really make sense, so I squashed them together. I also renamed the form to `ProviderForm` so it can be just used as a drop-in into other screens.

skateman commented:
All the dependencies are in, so this is ready for testing. To make parity verification easier, I'm temporarily keeping both forms available. 

@himdel @mzazrivec @ZitaNemeckova @h-kataria and anyone else who would like to click a little and/or has cloud provider credentials

chessbyte commented:
@skateman if this is ready for final review, should the WIP label be removed?

skateman commented:
@chessbyte I don't really want anyone to merge it with both forms rendered on the page, I don't want anyone to merge it before I remove the old one. But I don't want to remove the old one before it's approved because it's easier to test if you can see both versions at the same time.

Hyperkid123 commented:
@skateman I can see the light on the end of a tunnel! :partying_face:

skateman commented:
@Hyperkid123 is it bright enough to get a review from you? :wink:

Hyperkid123 commented:
@skateman the UI seems pretty straight forward to me and I don't see any issues with it. That said I am very out of touch with MIQ. From the DDF point of view, I give it :+1:.

skateman commented:
There are some tradeoffs regarding the form schemas:
* default zone can't be preselected as it can be renamed, so I made the field mandatory
* openstack api version has been defaulted to `v3` and it's required - suggested by @aufi 
* openstack amqp security protocol selector is gone because it was never implemented
* openstack API port became a required field, with a default value
* openstack RSA key is textual because we don't have a file upload component in DDFv1

Any other minor inconsistencies which I might forgot about are the product of a similar tradeoff. As they are not part of this PR (except `name`, `zone` and `type`) please address them in issues in the provider repos. You can find the schemas under a `params_for_create` method.

ZitaNemeckova commented:
@skateman I tried to add OpenStack one.

Request with following
```
{"action":"verify_credentials","resource":{"endpoints":{"amqp":{"hostname":"HOSTNAME","port":5672}},"authentications":{"amqp":{"userid":"USERID","password":"PASSWORD"}},"type":"ManageIQ::Providers::Openstack::CloudManager","event_stream_selection":"amqp"}}
```
returned:
```
{"results":[{"success":true,"message":"Credentials sent for verification","task_id":"59651","task_href":"http://localhost:3000/api/tasks/59651"}]}
```
But form shows:
<img width="266" alt="Screenshot 2020-05-14 at 14 46 52" src="https://user-images.githubusercontent.com/9210860/81936016-cd2a9b80-95f1-11ea-95a6-548653f77f14.png">

Validation was a success in old form.

skateman commented:
@ZitaNemeckova do you have a working AMQP endpoint? :open_mouth:

h-kataria commented:
@skateman DDF forms have Save/Reset/Cancel button on the Add provider form, those should be Add/Cancel buttons, Edit Provider forms should have Save/Reset/Cancel buttons.

h-kataria commented:
Adding Azure/Amazon Providers worked for me.

skateman commented:
@ZitaNemeckova the validation should fail in the old provider form, if I look at the log, I see:
```
Could not establish TCP connection to HOSTNAME
```
Which is consistent with what @aufi said about the AMQP not being available on our OpenStack provider. However, there's an issue with the port number not being passed to the validation, which I am addressing here: https://github.com/ManageIQ/manageiq-providers-openstack/pull/584

himdel commented on app/javascript/components/provider-form/index.jsx:
```diff
@@ -0,0 +1,185 @@
+import { pick, keyBy } from 'lodash';
+
+import { API } from '../../http_api';
+import MiqFormRenderer from '../../forms/data-driven-form';
```
`from @@ddf` ?

himdel commented:
Editing any provider:

* clicking Change on the password field gives...

```
Warning: React does not recognize the `FormSpyProvider` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `formspyprovider` instead. If you accidentally passed it from a parent component, remove it from the DOM element.
    in input (created by FormControl)
    in FormControl
    in span (created by InputGroup)
    in InputGroup
    in div (created by FormGroup)
    in FormGroup
    in Unknown (created by Field)
    in Field (created by Context.Consumer)
    in ReactFinalForm(Field) (created by FieldProvider)
    in FieldProvider (created by EditPasswordField)
    in EditPasswordField (created by FieldWrapper)
    in FieldWrapper (created by Context.Consumer)
    in FormFieldHideWrapper (created by Context.Consumer)
    in FormConditionWrapper (created by Context.Consumer)
    in PasswordField (created by FieldWrapper)
    in FieldWrapper (created by Context.Consumer)
    in FormFieldHideWrapper (created by Context.Consumer)
    in FormConditionWrapper (created by Context.Consumer)
    in AsyncCredentials (created by ValidateProviderCredentials)
    in ValidateProviderCredentials (created by FieldWrapper)
    in FieldWrapper (created by Context.Consumer)
    in FormFieldHideWrapper (created by Context.Consumer)
    in FormConditionWrapper (created by Context.Consumer)
    in div (created by SubForm)
    in SubForm (created by FieldWrapper)
    in FieldWrapper (created by Context.Consumer)
    in FormFieldHideWrapper (created by Context.Consumer)
    in FormConditionWrapper (created by Context.Consumer)
    in div (created by SubForm)
    in SubForm (created by FieldWrapper)
    in FieldWrapper (created by Context.Consumer)
    in FormFieldHideWrapper (created by Context.Consumer)
    in FormConditionWrapper (created by Context.Consumer)
    in form (created by Form)
    in Form (created by FormWrapper)
    in FormWrapper (created by Context.Consumer)
    in ReactFinalForm (created by FormRenderer)
    in FormRenderer (created by MiqFormRenderer)
    in MiqFormRenderer (created by ConnectFunction)
    in ConnectFunction (created by ProviderForm)
    in div (created by ProviderForm)
    in ProviderForm
    in Provider
```

.. but only the first time, change + cancel + change work without a 2nd warning


Editing an Amazon provider:

* Region is now editable, wasn't before


Editing an Azure provider:

* Region **and** Subscription ID are editable


Editing a Google provider:

* there is no Validate button when editing Service Account JSON


Editing an Openstack provider:

* works :) I assume the Private key file upload -> textarea difference is intentional?

skateman commented:
> Editing any provider:
> 
> * clicking Change on the password field gives...
> 
> ```
> Warning: React does not recognize the `FormSpyProvider` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `formspyprovider` instead. If you accidentally passed it from a parent component, remove it from the DOM element.
>     in input (created by FormControl)
>     in FormControl
>     in span (created by InputGroup)
>     in InputGroup
>     in div (created by FormGroup)
>     in FormGroup
>     in Unknown (created by Field)
>     in Field (created by Context.Consumer)
>     in ReactFinalForm(Field) (created by FieldProvider)
>     in FieldProvider (created by EditPasswordField)
>     in EditPasswordField (created by FieldWrapper)
>     in FieldWrapper (created by Context.Consumer)
>     in FormFieldHideWrapper (created by Context.Consumer)
>     in FormConditionWrapper (created by Context.Consumer)
>     in PasswordField (created by FieldWrapper)
>     in FieldWrapper (created by Context.Consumer)
>     in FormFieldHideWrapper (created by Context.Consumer)
>     in FormConditionWrapper (created by Context.Consumer)
>     in AsyncCredentials (created by ValidateProviderCredentials)
>     in ValidateProviderCredentials (created by FieldWrapper)
>     in FieldWrapper (created by Context.Consumer)
>     in FormFieldHideWrapper (created by Context.Consumer)
>     in FormConditionWrapper (created by Context.Consumer)
>     in div (created by SubForm)
>     in SubForm (created by FieldWrapper)
>     in FieldWrapper (created by Context.Consumer)
>     in FormFieldHideWrapper (created by Context.Consumer)
>     in FormConditionWrapper (created by Context.Consumer)
>     in div (created by SubForm)
>     in SubForm (created by FieldWrapper)
>     in FieldWrapper (created by Context.Consumer)
>     in FormFieldHideWrapper (created by Context.Consumer)
>     in FormConditionWrapper (created by Context.Consumer)
>     in form (created by Form)
>     in Form (created by FormWrapper)
>     in FormWrapper (created by Context.Consumer)
>     in ReactFinalForm (created by FormRenderer)
>     in FormRenderer (created by MiqFormRenderer)
>     in MiqFormRenderer (created by ConnectFunction)
>     in ConnectFunction (created by ProviderForm)
>     in div (created by ProviderForm)
>     in ProviderForm
>     in Provider
> ```
> 
> .. but only the first time, change + cancel + change work without a 2nd warning

DDFv2 will solve this

> 
> Editing an Amazon provider:
> 
> * Region is now editable, wasn't before

discussed it with @agrare, he said it's okay, backend support is there

> 
> Editing an Azure provider:
> 
> * Region **and** Subscription ID are editable
> 

same

> Editing a Google provider:
> 
> * there is no Validate button when editing Service Account JSON

is there one in the old version?

> 
> Editing an Openstack provider:
> 
> * works :) I assume the Private key file upload -> textarea difference is intentional?

yes!

himdel commented:
> > Google provider: there is no Validate button when editing Service Account JSON

> is there one in the old version?

![google](https://user-images.githubusercontent.com/289743/82948546-7ded4f00-9f91-11ea-8d33-d7daa993d004.png)

skateman commented:
> > > Google provider: there is no Validate button when editing Service Account JSON
> 
> > is there one in the old version?
> 
> ![google](https://user-images.githubusercontent.com/289743/82948546-7ded4f00-9f91-11ea-8d33-d7daa993d004.png)

It's still not related to this PR :laughing: can you create an issue in the google provider pls?

himdel commented:
Sure :) https://github.com/ManageIQ/manageiq-providers-google/issues/135

himdel commented:
When adding a new one..

![search](https://user-images.githubusercontent.com/289743/82951942-a24c2a00-9f97-11ea-92e0-5f62c5887dd7.png)

`Search...` is wrong as that select is not searchable.
But it's not in providers-amazon, it doesn't say `Search`, it seems to be a ddf default - maybe we should override the default to Choose? (globally in miq ddf configuration I mean)

himdel commented:
Also, when adding Openstack...

before, Openstack Infra Provider could be left to `<blank>`.
now, it starts with "Search...", and when I choose a provider, I can't un-choose it.

himdel commented on app/javascript/components/provider-form/index.jsx:
```diff
@@ -0,0 +1,185 @@
+};
+
+ProviderForm.propTypes = {
+  providerId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
```
should we even allow PropTypes.number?

himdel commented on app/views/ems_cloud/edit.html.haml:
```diff
@@ -1,3 +1,5 @@
+= react('ProviderForm', :providerId => @ems.id, :redirect => edit_redirect_path(@lastaction, @ems), :kind => 'cloud', :title => ui_lookup(:model => 'ManageIQ::Providers::CloudManager'))
```
`@ems.id.to_s`

skateman commented:
> Also, when adding Openstack...
> 
> before, Openstack Infra Provider could be left to `<blank>`.
> now, it starts with "Search...", and when I choose a provider, I can't un-choose it (yet it's not required).
> 
> ^ probably a problem for any non-required select?

This is also a provider repo issue, not related to this code. Will fix it there.

skateman commented:
> When adding a new one..
> 
> ![search](https://user-images.githubusercontent.com/289743/82951942-a24c2a00-9f97-11ea-92e0-5f62c5887dd7.png)
> 
> `Search...` is wrong as that select is not searchable.
> But it's not in providers-amazon, it doesn't say `Search`, it seems to be a ddf default - maybe we should override the default to Choose? (globally in miq ddf configuration I mean)

This is not related to this PR, also happens on master:
![Screenshot from 2020-05-27 09-20-25](https://user-images.githubusercontent.com/649130/82989695-60ed6600-9ffb-11ea-9555-cf41d6f5ca58.png)

This is what happens when you depend on a 3rd party library thta

Hyperkid123 commented:
@skateman you can use https://data-driven-forms.org/renderer/global-component-props if you want to override global props.

skateman commented:
> @skateman you can use https://data-driven-forms.org/renderer/global-component-props if you want to override global props.

That's great, but does this work on DDFv1?

Hyperkid123 commented:
@skateman oh right sorry. No its v2 feature. I keep forgetting.

h-kataria commented:
@skateman Add new Provider form should have "Add/Cancel" buttons instead of "Save/Reset/Cancel", to be consistent with other forms in the product

skateman commented:
> @skateman
> 
> * Add new Provider form should have "Add/Cancel" buttons instead of "Save/Reset/Cancel", to be consistent with other forms in the product

Fixed

> * Flash message is missing when after changes in the Edit provider form are Reset

I'm not entirely sure why is this necessary, the reset doesn't do a server roundtrip anymore so there's no way it would fail due to a server error.

> * Just wondering if form buttons should be disabled while Validation is in process on Edit provider form.

Maybe it's possible to do it, but I'd like to take it into a separate PR and especially after we are on DDFv2.

h-kataria commented:
> > @skateman
> > 
> > * Add new Provider form should have "Add/Cancel" buttons instead of "Save/Reset/Cancel", to be consistent with other forms in the product
> 
> Fixed
> 
> > * Flash message is missing when after changes in the Edit provider form are Reset
> 
> I'm not entirely sure why is this necessary, the reset doesn't do a server roundtrip anymore so there's no way it would fail due to a server error.
> 
> > * Just wondering if form buttons should be disabled while Validation is in process on Edit provider form.
> 
> Maybe it's possible to do it, but I'd like to take it into a separate PR and especially after we are on DDFv2.

I feel the Save should not be allowed while credentials are being validated, that will give user an opportunity to change/fix them in case validation failed. Let's do the disabling of Save button in a separate PR

I am not sure about removing the Reset flash message, we do have that in other DDF forms tho, is there a plan to remove it from all new forms?

skateman commented:
Okay, I'll put it there...but will create an issue about getting rid of this.
